src/logic/map_objects/tribes/ship.cc

```diff
@@ -190,7 +190,10 @@ void Ship::cleanup(EditorGameBase& egbase) {
 	}
 
 	while (!items_.empty()) {
-		items_.back().remove(egbase);
+		if (o != nullptr && ((ware_economy_ && o->has_economy(ware_economy_->serial())) ||
+		                     (worker_economy_ && o->has_economy(worker_economy_->serial())))) {
+			items_.back().remove(egbase);
```

Comment:
This looks like a potential memory leak to me, which is undesirable under all circumstances (*including* end-of-game cleanup). Did you check this with ASan?

Reply:
Checking for player was a bit silly, I agree. It is normal that ships outlive player. I checked with valgrind, but the test was quite limited as it is damn slow => no seen leaks does not translate to no leaks.

Reply:
This hid the problem in some cases. I guess that it would also be ok to remove all the new checks.

Reply:
If the economy is no longer present and ship is not empty, then what is the correct way to not leak memory? Calling remove() is bad, at least in case of workers; unsure of wares.

Reply:
I'll recommit after Travis finishes first round.